include/pybind11/pytypes.h

```diff
@@ -1638,6 +1638,13 @@ class capsule : public object {
     }
 
     const char *name() const { return PyCapsule_GetName(m_ptr); }
+
+    /// Replaces a capsule's name *without* calling the destructor on the existing one.
+    void set_name(const char *&new_name) {
```

Comment:
Why `const char *&new_name` instead of just `const char *`?
```suggestion
    void set_name(const char *new_name) {
```

Also: 
"If non-NULL, the name must outlive the capsule. If the previous name stored in the capsule was not NULL, no attempt is made to free it."

Shouldn't we try to free the old name? In our `set_name` function?

Reply:
No, usually the names are `static const char *`, so we should not attempt to free it.

Re: why `const char *&new_name` instead of `const char *` was to avoid copying the value of the pointer. If you think I am overthinking this, I am OK with using `const char *` instead.

Reply:
Any decent compiler should optimize this out.

The arg would still have to be copied for PyCapsule_SetName as that takes a `const char *` so this wouldn't be a useful optimization, it just makes the method overly complex.

Reply:
@oleksandr-pavlyk Actually, I just realized that the default destructors need to be changed for the explicit constructor since some of the make the assumption that the name of the capsule ("null") doesn't change, where not it could in those scenarios. They need to call name() for the calls to PyCapsule_GetPointer

Reply:
Or better yet, just call `PyCapsule_GetName`

Reply:
My thinking: Unless the function is inlined, the machine code will have to put *something* on the process stack. AFAIK it cannot get any better than a pointer. A reference is just a syntax trick, but ultimately it's still a pointer underneath.

But maybe more importantly: what if the capsule deleter manages the name, without storing the original char pointer elsewhere? — Obviously that cannot be solved here, the Python C API is fundamentally unsafe. But it might be good to hint on it in the comment. Trying:

/// Thin wrapper for PyCapsule_SetName. As with the Python C API call, it is the responsibility of the caller to ensure that the lifetime of the old and new pointees is managed properly.

Reply:
@rwgk The Capsule Deleter can just retrieve the name use PyCapsule_GetName(). Currently, we have it hard coded for two destructors we defined.